pkg/cloudprovider/aws/instance.go

```diff
@@ -185,10 +246,14 @@ func (p *InstanceProvider) getOverrides(instanceTypeOptions []cloudprovider.Inst
 				continue
 			}
 			for _, subnet := range subnets {
-				if aws.StringValue(subnet.AvailabilityZone) == offering.Zone {
+				subnetZone := aws.StringValue(subnet.AvailabilityZone)
+				if subnetZone == offering.Zone && (constrainedZones == nil || constrainedZones.Has(subnetZone)) {
```

Comment:
I think you're guaranteed that constrainedZones won't be nil, since Karpenter has to produce the full set of viable zones for the requirements `NotIn` operator to work. e.g. `us-west-2a` not in `nil` is impossible to compute.
```suggestion
				if zone := aws.StringValue(subnet.AvailabilityZone); zone == offering.Zone && zones.has(zone) {
```